quiz/quiz.6.in
.\"	$NetBSD: quiz.6,v 1.6 1997/09/20 14:28:16 lukem Exp $
.\"
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Jim R. Oldroyd at The Instruction Set.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)quiz.6	8.1 (Berkeley) 5/31/93
.\"
.Dd May 31, 1993
.Dt QUIZ 6
.Os
.Sh NAME
.Nm quiz
.Nd random knowledge tests
.Sh SYNOPSIS
.Nm
.Op Fl t
.Op Fl i Ar file
.Op Ar question answer
.Sh DESCRIPTION
The
.Nm
utility tests your knowledge of random facts.
It has a database of subjects from which you can choose.
With no arguments,
.Nm
displays the list of available subjects.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl t
Use tutorial mode, in which questions are repeated later if you didn't get
them right the first time, and new questions are presented less frequently
to help you learn the older ones.
.It Fl i
Specify an alternative index file.
.El
.Pp
Subjects are divided into categories.
You can pick any two categories from the same subject.
.Nm
will ask questions from the first category and it expects answers from
the second category.
For example, the command ``quiz victim killer'' asks questions which are
the names of victims, and expects you to answer with the cause of their
untimely demise, whereas the command ``quiz killer victim'' works the
other way around.
.Pp
If you get the answer wrong,
.Nm
lets you try again.
To see the right answer, enter a blank line.
.Sh "Index and Data File Syntax"
The index and data files have a similar syntax.
Lines in them consist of several categories separated by colons.
The categories are regular expressions formed using the following
meta-characters:
.sp
.Bl -tag -width "pat|pat" -compact -offset indent
.It pat|pat
alternative patterns
.It {pat}
optional pattern
.It [pat]
delimiters, as in pat[pat|pat]pat
.El
.Pp
In an index file, each line represents a subject.
The first category in each subject is the pathname of the data file for
the subject.
The remaining categories are regular expressions for the titles of each
category in the subject.
.Pp
In data files, each line represents a question/answer set.
Each category is the information for the question/answer for that category.
.Pp
The backslash character (``\e'') is used to quote syntactically significant
characters, or at the end of a line to signify that a continuation line
follows.
.Pp
If either a question or its answer is empty,
.Nm
will refrain from asking it.
.Sh FILES
.Bl -tag -width @quiz_dir@ -compact
.It Pa @quiz_dir@
The default index and data files.
.El
.Sh BUGS
.Nm
is pretty cynical about certain subjects.


NEWS
NEWS for bsd-games and bsd-games-non-free
=========================================

This file contains a summary of changes in each version of bsd-games
and bsd-games-non-free after 1.3.  For fuller details, see ChangeLog
and ChangeLog.0.

As well as reading this summary, it is a good idea to reread the
INSTALL file whenever installing a new version, to check for any
changes in installation instructions.  Packagers should also reread
the PACKAGING file each version (or do a diff between the old and new
versions of the package and read that).

Version 2.10 (2000-02-12)
============

* Avoid world-readable characs file in phantasia.  Note the new
configuration question which will need an appropriate answer,
especially if packaging for a distribution.

* Various minor bugfixes.

* Essentially all relevant changes, except to hunt (which will
eventually be replaced by the OpenBSD version), now merged back into
NetBSD.

* Some minor changes from NetBSD CVS of 2000-02-12.

* No new bsd-games-non-free version; bsd-games-non-free 2.8 is still
current.

Version 2.9 (1999-11-12)
===========

* Fixes various minor bugs reported recently.

* Some minor changes from NetBSD CVS of 1999-11-12.

* No new bsd-games-non-free version; bsd-games-non-free 2.8 is still
current.

Version 2.8 (1999-09-26)
===========

* Robots scorefile format has changed - remove or rename your old
score file when upgrading to this version.

* Various minor changes and bug-fixes (some from OpenBSD).

* Snake substantially cleaned up, and now uses the native curses
interface (rather than termcap emulation in ncurses); backgammon and
tetris still to be converted.

* New `game': wtf.

* Updated from NetBSD CVS of 1999-09-25; many more changes merged back
into NetBSD.

* All constant data files are now architecture independent (installed
under /usr/share by default).

* bsd-games is no longer tested by me with libc older than glibc
2.1.2; fixes for compilation with older libc will still be accepted.

* Details of the copyrights and licence terms applicable to these
games, as best as I can determine them, now appear in COPYING.

* bsd-games no longer comes with its own word list for boggle and
hangman; you need to provide one at compile time (for boggle) and at
run time (for hangman).  The copyright status of the one formerly
included was not entirely clear.  The word list used by hangman can
also now be specified at run time with the new `-d' option.

* bsd-games now has a testsuite (for programs that can be tested
non-interactively only).

Version 2.7 (1999-04-04)
===========

* Bug fixes.

* Updated from NetBSD-current of 1999-03-26; more changes merged back
into NetBSD.

* Other minor changes.

Version 2.6 (1998-12-20)
===========

* Buffer overrun fixes in sail (not thought exploitable).

* Fixes possible score file corruption if fds 0, 1, 2 aren't open at
game startup.

* Updated from NetBSD-current of 1998-12-07 (morse now has a decoding
option).

* Now builds cleanly with -Wwrite-strings.

* Other minor changes.

* Some users of this package may be interested in my `nppt' package,
which provides an alternative implementation of ppt(6) with additional
features such as Encapsulated PostScript output.  (This package
installs the binary as `nppt', so can be installed simultaneously with
this one.  Look for nppt-0.1.tar.gz (or a later version) on
metalab.unc.edu (formerly sunsite) in /pub/Linux/Incoming/ or under
/pub/Linux/games/.)

Version 2.5 (1998-09-24)
===========

* New build system, using a single project Makefile (autogenerated)
that includes makefile fragments from subdirectories, and uses GNU
Make's automatic dependency generation to generate dependencies
automatically, rather than recursive make.  See Peter Miller's paper
`Recursive Make Considered Harmful'
<http://www.canb.auug.org.au/~millerp/rmch/recu-make-cons-harm.html>
for reasons why this is better than recursive make.  This should make
no difference to normal users, but if you modify the sources note that
you will need to rebuild from the top directory, and you may find that
the dependencies are now accurate whereas formerly they were
incomplete.  The automatic dependency generation causes the build to
start with about 300 `no such file or directory' warnings, which
should be ignored: these refer to the dependency files which GNU Make
will then automatically generate.

* Should work with glibc 2.1 (at least, version 2.0.96 on x86, which
is what I've tested).  See the ChangeLog for an explanation of the
problem previous releases had with glibc 2.1.

* Boggle is now in the main bsd-games package - Paul Janzen checked
with the original author, who confirmed that the BSD licence
superseded the old non-free one.  bsd-games-non-free now only contains
rogue.

* Other minor changes.

Version 2.4 (1998-09-05)
===========

* Various minor bug fixes and improvements (some from OpenBSD, thanks
to Paul Janzen <pjanzen@foatdi.harvard.edu>).

* Sail directory can now be non-world-accessible (mode 2770 root.games
or similar).  This may improve security, since it means users cannot
tamper with sync files their sail process creates.

* Rogue buffer overrun fix.  An audit is still needed.

* dm now ignores the environment variable TZ when deciding at what
times games can be played.  (This only affects you if you are using dm
to control at what times games can be played, and not then if you are
using libc 5.4.45 or 5.4.46 (which completely ignore TZ in setgid
programs rather than checking the validity of any filename given as
glibc does).)

* All games updated from NetBSD-current of 1998-08-30 (minor changes;
some more of my changes have been merged back into NetBSD).

Version 2.3 (1998-08-19)
===========

* A possible temporary file security bug in sail fixed.

* Countmail is back, with a licence (apparently the standard NetBSD
one), since it has one in NetBSD-current of 1998-08-19.

* All the other games updated from NetBSD-current of 1998-08-19 (minor
changes).


quiz/quiz.c
/*	$NetBSD: quiz.c,v 1.19 2003/08/07 09:37:34 agc Exp $	*/

/*-
 * Copyright (c) 1991, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Jim R. Oldroyd at The Instruction Set and Keith Gabryelski at
 * Commodore Business Machines.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1991, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)quiz.c	8.3 (Berkeley) 5/4/95";
#else
__RCSID("$NetBSD: quiz.c,v 1.19 2003/08/07 09:37:34 agc Exp $");
#endif
#endif /* not lint */

#include <sys/types.h>

#include <ctype.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <err.h>
#include <time.h>
#include <unistd.h>
#include "quiz.h"
#include "pathnames.h"

static QE qlist;
static int catone, cattwo, tflag;
static u_int qsize;

char	*appdstr __P((char *, const char *, size_t));
void	 downcase __P((char *));
void	 get_cats __P((char *, char *));
void	 get_file __P((const char *));
int	 main __P((int, char *[]));
const char	*next_cat __P((const char *));
void	 quiz __P((void));
void	 score __P((u_int, u_int, u_int));
void	 show_index __P((void));
void	 usage __P((void)) __attribute__((__noreturn__));

int
main(argc, argv)
	int argc;
	char *argv[];
{
	int ch;
	const char *indexfile;

	/* Revoke setgid privileges */
	setregid(getgid(), getgid());

	indexfile = _PATH_QUIZIDX;
	while ((ch = getopt(argc, argv, "i:t")) != -1)
		switch(ch) {
		case 'i':
			indexfile = optarg;
			break;
		case 't':
			tflag = 1;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	switch(argc) {
	case 0:
		get_file(indexfile);
		show_index();
		break;
	case 2:
		get_file(indexfile);
		get_cats(argv[0], argv[1]);
		quiz();
		break;
	default:
		usage();
	}
	exit(0);
}

void
get_file(file)
	const char *file;
{
	FILE *fp;
	QE *qp;
	size_t len;
	char *lp;

	if ((fp = fopen(file, "r")) == NULL)
		err(1, "%s", file);

	/*
	 * XXX
	 * Should really free up space from any earlier read list
	 * but there are no reverse pointers to do so with.
	 */
	qp = &qlist;
	qsize = 0;
	while ((lp = fgetln(fp, &len)) != NULL) {
		if (lp[len - 1] == '\n')
			lp[--len] = '\0';
		if (qp->q_text && qp->q_text[strlen(qp->q_text) - 1] == '\\')
			qp->q_text = appdstr(qp->q_text, lp, len);
		else {
			if ((qp->q_next = malloc(sizeof(QE))) == NULL)
				errx(1, "malloc");
			qp = qp->q_next;
			if ((qp->q_text = malloc(len + 1)) == NULL)
				errx(1, "malloc");
			strncpy(qp->q_text, lp, len);
			qp->q_text[len] = '\0';
			qp->q_asked = qp->q_answered = FALSE;
			qp->q_next = NULL;
			++qsize;
		}
	}
	(void)fclose(fp);
}

void
show_index()
{
	QE *qp;
	const char *p, *s;
	FILE *pf;
	const char *pager;

	if (!isatty(1))
		pager = "cat";
	else {
		if (!(pager = getenv("PAGER")) || (*pager == 0))
			pager = _PATH_PAGER;
	}
	if ((pf = popen(pager, "w")) == NULL)
		err(1, "%s", pager);
	(void)fprintf(pf, "Subjects:\n\n");
	for (qp = qlist.q_next; qp; qp = qp->q_next) {
		for (s = next_cat(qp->q_text); s; s = next_cat(s)) {
			if (!rxp_compile(s))
				errx(1, "%s", rxperr);
			if ((p = rxp_expand()) != NULL)
				(void)fprintf(pf, "%s ", p);
		}
		(void)fprintf(pf, "\n");
	}
	(void)fprintf(pf, "\n%s\n%s\n%s\n",
"For example, \"quiz victim killer\" prints a victim's name and you reply",
"with the killer, and \"quiz killer victim\" works the other way around.",
"Type an empty line to get the correct answer.");
	(void)pclose(pf);
}

void
get_cats(cat1, cat2)
	char *cat1, *cat2;
{
	QE *qp;
	int i;
	const char *s;

	downcase(cat1);
	downcase(cat2);
	for (qp = qlist.q_next; qp; qp = qp->q_next) {
		s = next_cat(qp->q_text);
		catone = cattwo = i = 0;
		while (s) {
			if (!rxp_compile(s))
				errx(1, "%s", rxperr);
			i++;
			if (rxp_match(cat1))
				catone = i;
			if (rxp_match(cat2))
				cattwo = i;
			s = next_cat(s);
		}
		if (catone && cattwo && catone != cattwo) {
			if (!rxp_compile(qp->q_text))
				errx(1, "%s", rxperr);
			get_file(rxp_expand());
			return;
		}
	}
	errx(1, "invalid categories");
}

void
quiz()
{
	QE *qp;
	int i;
	size_t len;
	u_int guesses, rights, wrongs;
	int next;
	char *answer, *t, question[LINE_SZ];
	const char *s;

	srandom(time(NULL));
	guesses = rights = wrongs = 0;
	for (;;) {
		if (qsize == 0)
			break;
		next = random() % qsize;
		qp = qlist.q_next;
		for (i = 0; i < next; i++)
			qp = qp->q_next;
		while (qp && qp->q_answered)
			qp = qp->q_next;
		if (!qp) {
			qsize = next;
			continue;
		}
		if (tflag && random() % 100 > 20) {
			/* repeat questions in tutorial mode */
			while (qp && (!qp->q_asked || qp->q_answered))
				qp = qp->q_next;
			if (!qp)
				continue;
		}
		s = qp->q_text;
		for (i = 0; i < catone - 1; i++)
			s = next_cat(s);
		if (!rxp_compile(s))
			errx(1, "%s", rxperr);
		t = rxp_expand();
		if (!t || *t == '\0') {
			qp->q_answered = TRUE;
			continue;
		}
		(void)strcpy(question, t);
		s = qp->q_text;
		for (i = 0; i < cattwo - 1; i++)
			s = next_cat(s);
		if (!rxp_compile(s))
			errx(1, "%s", rxperr);
		t = rxp_expand();
		if (!t || *t == '\0') {
			qp->q_answered = TRUE;
			continue;
		}
		qp->q_asked = TRUE;
		(void)printf("%s?\n", question);
		for (;; ++guesses) {
			if ((answer = fgetln(stdin, &len)) == NULL ||
			    answer[len - 1] != '\n') {
				score(rights, wrongs, guesses);
				exit(0);
			}
			answer[len - 1] = '\0';
			downcase(answer);
			if (rxp_match(answer)) {
				(void)printf("Right!\n");
				++rights;
				qp->q_answered = TRUE;
				break;
			}
			if (*answer == '\0') {
				(void)printf("%s\n", t);
				++wrongs;
				if (!tflag)
					qp->q_answered = TRUE;
				break;
			}
			(void)printf("What?\n");
		}
	}
	score(rights, wrongs, guesses);
}

const char *
next_cat(s)
	const char *	s;
{
	int esc;

	esc = 0;
	for (;;)
		switch (*s++) {
		case '\0':
			return (NULL);
		case '\\':
			esc = 1;
			break;
		case ':':
			if (!esc)
				return (s);
		default:
			esc = 0;
			break;
		}
	/* NOTREACHED */
}

char *
appdstr(s, tp, len)
	char *s;
	const char *tp;
	size_t len;
{
	char *mp;
	const char *sp;
	int ch;
	char *m;

	if ((m = malloc(strlen(s) + len + 1)) == NULL)
		errx(1, "malloc");
	for (mp = m, sp = s; (*mp++ = *sp++) != '\0'; )
		;
	--mp;
	if (*(mp - 1) == '\\')
		--mp;

	while ((ch = *mp++ = *tp++) && ch != '\n')
		;
	*mp = '\0';

	free(s);
	return (m);
}

void
score(r, w, g)
	u_int r, w, g;
{
	(void)printf("Rights %d, wrongs %d,", r, w);
	if (g)
		(void)printf(" extra guesses %d,", g);
	(void)printf(" score %d%%\n", (r + w + g) ? r * 100 / (r + w + g) : 0);
}

void
downcase(p)
	char *p;
{
	int ch;

	for (; (ch = *p) != '\0'; ++p)
		if (isascii(ch) && isupper(ch))
			*p = tolower(ch);
}

void
usage()
{
	(void)fprintf(stderr, "quiz [-t] [-i file] category1 category2\n");
	exit(1);
}
